File: faker-egy-names-php/src/Quality.php
<?php

declare(strict_types=1);

namespace Afify\FakerEgyNames;

/**
 * Personal-name quality gate — same semantics as php/egy-names's Quality,
 * python/src/egy_names/_quality.py's is_generatable_entry, and the
 * generate()-side filter in every other egy-names SDK.
 *
 * The generate catalog otherwise contains a handful of non-personal
 * tokens ("الله", "البيت", ...), malformed compounds ("د الدين", glued
 * three-name chains), and low-confidence filler rows. None of those
 * are a real person's name and generate() must never emit them.
 */
final class Quality
{
    /** @var array<string, mixed>|null */
    private static ?array $config = null;

    /** @var array<string, true>|null keyed by 'a' across the whole catalog */
    private static ?array $arLookup = null;

    /**
     * @param list<array{a: string, e: string, g: string, r: string, l: string, fc: string, ma?: string, p: list<float>, tp: float}> $entry
     */
    public static function isGeneratableEntry(array $entry): bool
    {
        $ar = trim($entry['a']);

        return !self::isNonPersonal($ar)
            && !self::isLowConfidence($entry)
            && !str_contains($ar, ' ');
    }

    private static function isNonPersonal(string $ar): bool
    {
        return in_array($ar, self::config()['quality']['non_personal_ar'], true);
    }

    /**
     * @param array{a: string, ma?: string, tp: float} $entry
     */
    private static function isLowConfidence(array $entry): bool
    {
        if (self::isMalformedCompound($entry)) {
            return true;
        }
        $epsilon = self::config()['quality']['low_confidence_share_epsilon'];
        if (($entry['tp'] ?? 0.0) > $epsilon) {
            return false;
        }
        $meaning = $entry['ma'] ?? '';
        foreach (self::config()['quality']['uncertain_meaning_markers'] as $marker) {
            if ($marker !== '' && str_contains($meaning, $marker)) {
                return true;
            }
        }
        return false;
    }

    /**
     * True if a multi-word lemma's first half is not a real name.
     *
     * Well-formed two-word lemmas are either a kunya ("أبو" + element) or
     * a name plus a compound. Everything else at the corpus noise floor
     * whose first token does not independently resolve is a corrupted row.
     *
     * @param array{a: string, tp: float} $entry
     */
    private static function isMalformedCompound(array $entry): bool
    {
        $ar = trim($entry['a']);
        if (!str_contains($ar, ' ')) {
            return false;
        }
        $epsilon = self::config()['quality']['low_confidence_share_epsilon'];
        if (($entry['tp'] ?? 0.0) > $epsilon) {
            return false;
        }

        $parts = preg_split('/\s+/u', $ar) ?: [];
        $first = $parts[0] ?? '';

        if (in_array($first, self::config()['quality']['kunya_exempt_prefixes'], true)) {
            return false;
        }

        return !isset(self::arLookup()[$first]);
    }

    /**
     * @return array<string, true>
     */
    private static function arLookup(): array
    {
        if (self::$arLookup === null) {
            self::$arLookup = [];
            foreach (Catalog::names() as $entry) {
                self::$arLookup[trim($entry['a'])] = true;
            }
        }
        return self::$arLookup;
    }

    /**
     * @return array<string, mixed>
     */
    private static function config(): array
    {
        if (self::$config !== null) {
            return self::$config;
        }

        $fallback = [
            'quality' => [
                'non_personal_ar' => [
                    'الله', 'الرجل', 'الرجال', 'شربه', 'لافندي', 'لفندي', 'ماء', 'البيت',
                ],
                'uncertain_meaning_markers' => [
                    'غير واضح', 'لا يوجد معنى', 'غير معروف',
                    'قد يكون تحريف', 'تحريفاً', 'تحريفًا',
                ],
                'low_confidence_share_epsilon' => 0.0001,
                'kunya_exempt_prefixes' => ['أبو', 'ابو', 'أم', 'ام'],
            ],
        ];

        $path = dirname(__DIR__) . '/data/logic_config.json';
        $raw = @file_get_contents($path);
        if ($raw === false) {
            self::$config = $fallback;
            return self::$config;
        }

        $decoded = json_decode($raw, true);
        if (!is_array($decoded) || !isset($decoded['quality']) || !is_array($decoded['quality'])) {
            self::$config = $fallback;
            return self::$config;
        }

        self::$config = $decoded;
        return self::$config;
    }
}

File: java/egy-names/src/main/java/com/afify/egynames/index/Quality.java
package com.afify.egynames.index;

import com.afify.egynames.model.Models;

import java.util.List;
import java.util.Set;

/**
 * Personal-name quality gate.
 *
 * <p>The book keeps some surface tokens so split and compounds still resolve.
 * Those tokens are not a person's name. Production APIs must not treat them
 * as one. Mirrors {@code python/src/egy_names/_quality.py}.
 */
public final class Quality {

    private Quality() {}

    /** True if this lemma may stand as a person's name. */
    public static boolean isPersonalEntry(Models.NameEntry entry) {
        return entry != null && !RulesConfig.nonPersonalAr().contains(entry.ar);
    }

    /**
     * True if this lemma is likely a fabricated/unverified filler row.
     *
     * <p>Zero real-world corpus share on its own just means "rare" — plenty
     * of genuine names are rare. But zero share combined with the catalog's
     * own gloss admitting the name's meaning/origin is unclear is a strong
     * signal the row was never attested, only guessed.
     */
    public static boolean isLowConfidenceEntry(Models.NameEntry entry) {
        if (entry == null) return true;
        if (isMalformedCompound(entry)) return true;
        if (entry.corpusShare > RulesConfig.lowConfidenceShareEpsilon()) return false;
        String meaning = entry.meaningAr != null ? entry.meaningAr : "";
        for (String marker : RulesConfig.uncertainMeaningMarkers()) {
            if (meaning.contains(marker)) return true;
        }
        return false;
    }

    /**
     * True if a multi-word lemma's first half is not a real name.
     *
     * <p>A well-formed two-word lemma is either a kunya ("أبو" + element) or
     * a name plus a compound. What is left after spacing fixes are corrupted
     * rows whose first half resolves to nothing, and they all sit at the
     * corpus noise floor.
     */
    public static boolean isMalformedCompound(Models.NameEntry entry) {
        if (entry == null || entry.ar == null) return false;
        String ar = entry.ar.trim();
        if (!ar.contains(" ")) return false;
        if (entry.corpusShare > RulesConfig.lowConfidenceShareEpsilon()) return false;
        String[] words = ar.split("\\s+");
        String first = words.length > 0 ? words[0] : "";
        if (RulesConfig.kunyaExemptPrefixes().contains(first)) return false;
        return LookupIndices.lookupAr(first, null) == null;
    }

    /** True if generate may emit this lemma as one token. */
    public static boolean isGeneratableEntry(Models.NameEntry entry) {
        return entry != null
                && isPersonalEntry(entry)
                && !isLowConfidenceEntry(entry)
                && !entry.ar.trim().contains(" ");
    }

    /** Family and tribal tokens are lineage, not the person. */
    public static boolean isLineageRole(Models.NameEntry entry) {
        return entry != null && entry.role == Models.NameRole.FAMILY;
    }
}
